core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/timers/SyncTimeoutTask.java

```diff
@@ -38,18 +43,51 @@ public void abortable(Abortable abortable) {
         this.abortable = abortable;
     }
 
+    /**
+     * Runs this task. If cancel() was called prior to this invocation, has no side effects. Otherwise, behaves with the
+     * following post-conditions: (1) threadToInterrupt's interrupt flag is set to true (unless a concurrent process
+     * clears it); (2) hasExecuted() will return true.
+     *
+     * Note that run(), cancel(), and hasExecuted() behave atomically - calls to these methods operate with strict
+     * happens-before relationships to one another.
+     */
     @Override
     public void run() {
-        hasExecuted = true;
-        threadToInterrupt.interrupt();
+        synchronized (this.lock) {
+            if (isCancelled) {
+                return;
+            }
+            hasExecuted = true;
+            threadToInterrupt.interrupt();
+
+            if (abortable != null) {
+                abortable.abort();
+            }
+        }
+    }
 
-        if (abortable != null) {
-            abortable.abort();
+    // XXX: I'm pretty sure this does't actually have to be synchronized - just doing the volatile write should be
```

Comment:
I think I'll actually leave this in here. Without the lock, it's possible that when `run` takes the lock, `cancel()` is called right after it checks `isCancelled` is `false` so that at the end of `run` `isCancelled` and `hasExecuted` are both true.

I'll remove the comment for now.